### PR TITLE
Grpc ruby switch to new call api

### LINKED ISSUE #118: Replace Event wrapper with native object
E.g, return all the information in the event as a Struct.

### LINKED ISSUE #547: Move Ruby to the new core batch API
This is a tracking bug so I know when things are done

### LINKED ISSUE #770: Re-enable some ruby unit tests that have been skipped because c-layer assertions started to fail in VM shutdown
a couple of tests in active_call_spec.rb started to fail on an assertion during VM shutdown.
All other ruby unit tests are still fine, these tests are doing something unusual in that they create a Call instance but never actually use it, that's likely why they have been affected and the other tests and interop behaviour is still OK.

Here is the stack trace from the core dump.

(gdb) bt
#0  0x00007fd275da5cc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007fd275da90d8 in __GI_abort () at abort.c:89
#2  0x00007fd2715c6f0d in gpr_mu_lock () from /usr/local/lib/libgpr.so.0
#3  0x00007fd271832582 in grpc_cq_end_finished () from /usr/local/lib/libgrpc.so.0
#4  0x00007fd2718306fa in ?? () from /usr/local/lib/libgrpc.so.0
#5  0x00007fd271830040 in ?? () from /usr/local/lib/libgrpc.so.0
#6  0x00007fd27183091a in grpc_call_stream_closed () from /usr/local/lib/libgrpc.so.0
#7  0x00007fd271821584 in ?? () from /usr/local/lib/libgrpc.so.0
#8  0x00007fd271821a8e in ?? () from /usr/local/lib/libgrpc.so.0
#9  0x00007fd27182fecb in grpc_call_cancel () from /usr/local/lib/libgrpc.so.0
#10 0x00007fd271830550 in grpc_call_destroy () from /usr/local/lib/libgrpc.so.0
#11 0x00007fd2761cbb9f in run_final (objspace=0x6479a0, objspace=0x6479a0, obj=29541560) at gc.c:2064
#12 finalize_list (objspace=objspace@entry=0x6479a0, p=0x1c2c4b8) at gc.c:2080
#13 0x00007fd2761d7b0f in rb_objspace_call_finalizer (objspace=0x6479a0) at gc.c:2212
#14 rb_gc_call_finalizer_at_exit () at gc.c:2146
#15 0x00007fd2761b6d0f in ruby_finalize_1 () at eval.c:129
#16 ruby_cleanup (ex=0) at eval.c:230
#17 0x00007fd2761b6f43 in ruby_run_node (n=0xa84e68) at eval.c:310
#18 0x000000000040087b in main (argc=4, argv=0x7fffb4e0f6b8) at main.c:36

### PR DESCRIPTION
**Pull request by tbetbetbe**
Closes #770
Closes #547 
Closes #118 

There will be more related cleanup in a follow-up PR; this one is big enough as-is.

### CODE REVIEW

File: src/ruby/ext/grpc/rb_call.c [hunk 1]
```diff
@@ -299,147 +181,397 @@ static VALUE grpc_rb_call_set_metadata(VALUE self, VALUE metadata) {
...
-   Produces no events */
-static VALUE grpc_rb_call_server_end_initial_metadata(int argc, VALUE *argv,
-                                                      VALUE self) {
-  VALUE flags = Qnil;
-  grpc_call *call = NULL;
-  grpc_call_error err;
+/* grpc_run_batch_stack_cleanup ensures the run_batch_stack is properly
+ * cleaned up */
+static void grpc_run_batch_stack_cleanup(run_batch_stack* st) {
+  grpc_metadata_array_destroy(&st->send_metadata);
```
Comment (jtattermusch) [inline src/ruby/ext/grpc/rb_call.c]: Depending on the use case of metadata(e.g. sendMetadata vs recvMetadata), you need to also delete the keys and values of the metadata, so having the same handling for send_metadata and recv_metadata here looks suspicious.. Perhaps you are already doing the right cleanup that somewhere else, but please make sure you free everything that needs to be freed.

Comment (jtattermusch) [inline src/ruby/ext/grpc/rb_call.c]: Here's an example of what C# does.
https://github.com/grpc/grpc/blob/master/src/csharp/ext/grpc_csharp_ext.c#L194

File: src/ruby/ext/grpc/rb_call.c [hunk 2]
```diff
@@ -299,147 +181,397 @@ static VALUE grpc_rb_call_set_metadata(VALUE self, VALUE metadata) {
...
+  size_t i = 0;
+  VALUE result = rb_struct_new(rb_sBatchResult, Qnil, Qnil, Qnil, Qnil, Qnil,
+                               Qnil, Qnil, Qnil, NULL);
+  for (i = 0; i < st->op_num; i++) {
+    switch(st->ops[i].op) {
+      case GRPC_OP_SEND_INITIAL_METADATA:
+        rb_struct_aset(result, sym_send_metadata, Qtrue);
+        break;
+      case GRPC_OP_SEND_MESSAGE:
+        rb_struct_aset(result, sym_send_message, Qtrue);
```
Comment (jtattermusch) [inline src/ruby/ext/grpc/rb_call.c]: The send_message byte buffer should be freed here.

**Comment by murgatroid99** [comment]
Why did you close this?

File: src/ruby/ext/grpc/rb_call.c (hunk 1 above)
Comment (tbetbetbe) [inline src/ruby/ext/grpc/rb_call.c]: I looked into this, and I'm sure this is OK, at least its consistent with the C# example

On sending, the keys and values pointers to data that's already in ruby objects, and these will be cleaned up once those objects are garbage collected, so it's only necessary to cleanup the metadata.
On receiving, it's only necessary to clean up the metadata as the underlying c-layer will take care of cleaning the keys and values.  Ruby makes copies of them which it manages.

File: src/ruby/ext/grpc/rb_call.c (hunk 2 above)
Comment (tbetbetbe) [inline src/ruby/ext/grpc/rb_call.c]: Done

File: src/ruby/ext/grpc/rb_call.c [hunk 3]
```diff
@@ -299,147 +181,402 @@ static VALUE grpc_rb_call_set_metadata(VALUE self, VALUE metadata) {
...
+        grpc_byte_buffer_destroy(st->ops[st->op_num].data.send_message);
```
Comment (jtattermusch) [inline src/ruby/ext/grpc/rb_call.c]: st->ops[i] :-)

Comment (jtattermusch) [inline src/ruby/ext/grpc/rb_call.c]: Which makes me wonder how come the ruby tests don't fail. Can you rerun them?

Comment (tbetbetbe) [inline src/ruby/ext/grpc/rb_call.c]: Fixed.

**Pull request merged**